File: lib/services/database_service.dart
import 'dart:convert';
import 'dart:io';

import 'package:sqflite_common_ffi/sqflite_ffi.dart';
import 'package:path/path.dart';
import 'package:bugaoshan/models/course.dart';

const String _keyCurrentScheduleId = 'currentScheduleId';

class DatabaseService {
  late Database _db;

  // In-memory cache to keep synchronous getters working
  String _currentScheduleId = 'default';
  List<ScheduleConfig> _schedules = [];
  final Map<String, Course> _coursesCache = {}; // ID to Course mapping for the current schedule

  Future<void> init() async {
    if (Platform.isWindows || Platform.isLinux || Platform.isMacOS) {
      sqfliteFfiInit();
      databaseFactory = databaseFactoryFfi;
    }

    final databasesPath = await getDatabasesPath();
    final path = join(databasesPath, 'bugaoshan.db');

    _db = await openDatabase(
      path,
      version: 1,
      onCreate: (db, version) async {
        await db.execute('''
          CREATE TABLE metadata (
            key TEXT PRIMARY KEY,
            value TEXT
          )
        ''');
        await db.execute('''
          CREATE TABLE schedules (
            id TEXT PRIMARY KEY,
            data_json TEXT
          )
        ''');
        await db.execute('''
          CREATE TABLE courses (
            id TEXT PRIMARY KEY,
            schedule_id TEXT,
            data_json TEXT
          )
        ''');
      },
    );

    // Initialize Metadata
    final metaResult = await _db.query('metadata', where: 'key = ?', whereArgs: [_keyCurrentScheduleId]);
    if (metaResult.isNotEmpty) {
      _currentScheduleId = metaResult.first['value'] as String;
    } else {
      await _db.insert('metadata', {'key': _keyCurrentScheduleId, 'value': 'default'});
      _currentScheduleId = 'default';
    }

    // Load Schedules
    final schedulesResult = await _db.query('schedules');
    if (schedulesResult.isNotEmpty) {
      _schedules = schedulesResult.map((e) => ScheduleConfig.fromJson(_decodeJson(e['data_json'] as String))).toList();
    } else {
      // Create default
      final defaultConfig = _defaultScheduleConfig();
      _schedules = [defaultConfig];
      await _db.insert('schedules', {'id': defaultConfig.id, 'data_json': _encodeJson(defaultConfig.toJson())});
    }

    // Load Courses for current schedule
    await _loadCoursesCache(_currentScheduleId);
  }

  Future<void> _loadCoursesCache(String scheduleId) async {
    _coursesCache.clear();
    final coursesResult = await _db.query('courses', where: 'schedule_id = ?', whereArgs: [scheduleId]);
    for (var row in coursesResult) {
      final course = Course.fromJson(_decodeJson(row['data_json'] as String));
      _coursesCache[course.id] = course;
    }
  }

  // ==================== Schedules Management ====================

  String getCurrentScheduleId() {
    return _currentScheduleId;
  }

  Future<void> switchSchedule(String scheduleId) async {
    _currentScheduleId = scheduleId;
    await _db.update('metadata', {'value': scheduleId}, where: 'key = ?', whereArgs: [_keyCurrentScheduleId]);
    await _loadCoursesCache(scheduleId);
  }

  List<ScheduleConfig> getAllSchedules() {
    return List.unmodifiable(_schedules);
  }

  ScheduleConfig getScheduleConfig() {
    return _schedules.firstWhere(
      (s) => s.id == _currentScheduleId,
      orElse: () => _schedules.first,
    );
  }

  Future<void> saveScheduleConfig(ScheduleConfig config) async {
    final index = _schedules.indexWhere((s) => s.id == config.id);
    if (index >= 0) {
      _schedules[index] = config;
      await _db.update('schedules', {'data_json': _encodeJson(config.toJson())}, where: 'id = ?', whereArgs: [config.id]);
    } else {
      _schedules.add(config);
      await _db.insert('schedules', {'id': config.id, 'data_json': _encodeJson(config.toJson())});
    }
  }

  Future<void> addSchedule(ScheduleConfig config) async {
    await saveScheduleConfig(config);
  }

  Future<void> deleteSchedule(String scheduleId) async {
    _schedules.removeWhere((s) => s.id == scheduleId);
    await _db.delete('schedules', where: 'id = ?', whereArgs: [scheduleId]);
    await _db.delete('courses', where: 'schedule_id = ?', whereArgs: [scheduleId]);

    // If we deleted the current one, switch to the first available
    if (_currentScheduleId == scheduleId && _schedules.isNotEmpty) {
      await switchSchedule(_schedules.first.id);
    }
  }

  // ==================== Courses ====================

  List<Course> getCourses({String? scheduleId}) {
    // Synchronous call requires cache. We only cache the current schedule.
    if (scheduleId == null || scheduleId == _currentScheduleId) {
      return _coursesCache.values.toList();
    }
    // If they asked for a different schedule synchronously, we can't fetch it instantly.
    // In practice, this codebase only fetches the current one synchronously.
    return [];
  }

  Future<void> addCourse(Course course) async {
    _coursesCache[course.id] = course;
    await _db.insert('courses', {'id': course.id, 'schedule_id': _currentScheduleId, 'data_json': _encodeJson(course.toJson())});
  }

  Future<void> updateCourse(Course course) async {
    _coursesCache[course.id] = course;
    await _db.update('courses', {'data_json': _encodeJson(course.toJson())}, where: 'id = ?', whereArgs: [course.id]);
  }

  Future<void> deleteCourse(String courseId) async {
    _coursesCache.remove(courseId);
    await _db.delete('courses', where: 'id = ?', whereArgs: [courseId]);
  }

  Future<List<Course>> getCoursesAsync({String? scheduleId}) async {
    if (scheduleId == null || scheduleId == _currentScheduleId) {
      return _coursesCache.values.toList();
    }

    final coursesResult = await _db.query('courses', where: 'schedule_id = ?', whereArgs: [scheduleId]);
    return coursesResult.map((row) => Course.fromJson(_decodeJson(row['data_json'] as String))).toList();
  }

  Future<bool> hasConflict(Course course, {String? excludeId}) async {
    return getCourses().any(
      (c) => c.conflictsWith(course, excludeId: excludeId),
    );
  }

  // ==================== Clear All ====================

  Future<void> clearAllCourseData() async {
    _coursesCache.clear();
    await _db.delete('courses', where: 'schedule_id = ?', whereArgs: [_currentScheduleId]);
  }

  // ==================== Helpers ====================

  ScheduleConfig _defaultScheduleConfig() {
    final now = DateTime.now();
    return ScheduleConfig(
      id: 'default',
      semesterName: '默认课表',
      semesterStartDate: now.toMonday(),
      totalWeeks: 20,
    );
  }

  Map<String, dynamic> _decodeJson(String str) =>
      Map<String, dynamic>.from(json.decode(str) as Map);

  String _encodeJson(Map<String, dynamic> map) => json.encode(map);
}

File: shared/src/commonMain/kotlin/io/github/the_brotherhood_of_scu/bugaoshan/db/DatabaseService.kt
package io.github.the_brotherhood_of_scu.bugaoshan.db

import io.github.the_brotherhood_of_scu.bugaoshan.model.Course
import io.github.the_brotherhood_of_scu.bugaoshan.model.ScheduleConfig
import io.github.the_brotherhood_of_scu.bugaoshan.model.WeekType
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

class DatabaseService(private val database: AppDatabase) {
    private val json = Json { ignoreUnknownKeys = true }

    // In-memory cache
    private var currentScheduleId: String = "default"
    private var schedulesCache: List<ScheduleConfig> = emptyList()
    private var coursesCache: List<Course> = emptyList()

    suspend fun init() = withContext(Dispatchers.Default) {
        // Load current schedule ID
        val metaRow = database.metadataQueries.getMetadata(KEY_CURRENT_SCHEDULE_ID).executeAsOneOrNull()
        if (metaRow != null) {
            currentScheduleId = metaRow
        }

        // Ensure default schedule exists
        val scheduleCount = database.schedulesQueries.getAllSchedules().executeAsList()
        if (scheduleCount.isEmpty()) {
            val defaultConfig = ScheduleConfig.defaultConfig()
            database.schedulesQueries.insertSchedule(
                defaultConfig.id,
                json.encodeToString(defaultConfig)
            )
            database.metadataQueries.setMetadata(KEY_CURRENT_SCHEDULE_ID, defaultConfig.id)
        }

        // Load caches
        loadSchedulesCache()
        loadCoursesCache()
    }

    private fun loadSchedulesCache() {
        val rows = database.schedulesQueries.getAllSchedules().executeAsList()
        schedulesCache = rows.map { row ->
            json.decodeFromString<ScheduleConfig>(row.config_json)
        }
    }

    private fun loadCoursesCache() {
        val rows = database.coursesQueries.getCoursesBySchedule(currentScheduleId).executeAsList()
        coursesCache = rows.map { row ->
            Course(
                id = row.id,
                name = row.name ?: "",
                teacher = row.teacher ?: "",
                location = row.location ?: "",
                startWeek = row.start_week?.toInt() ?: 1,
                endWeek = row.end_week?.toInt() ?: 20,
                dayOfWeek = row.day_of_week?.toInt() ?: 1,
                startSection = row.start_section?.toInt() ?: 1,
                endSection = row.end_section?.toInt() ?: 2,
                colorValue = row.color_value ?: 0xFF2196F3,
                weekType = WeekType.fromIndex(row.week_type?.toInt() ?: 0),
            )
        }
    }

    // ==================== Schedule Management ====================

    fun getCurrentScheduleId(): String = currentScheduleId

    suspend fun switchSchedule(scheduleId: String) = withContext(Dispatchers.Default) {
        currentScheduleId = scheduleId
        database.metadataQueries.setMetadata(KEY_CURRENT_SCHEDULE_ID, scheduleId)
        loadCoursesCache()
    }

    fun getAllSchedules(): List<ScheduleConfig> = schedulesCache

    fun getScheduleConfig(): ScheduleConfig {
        return schedulesCache.firstOrNull { it.id == currentScheduleId }
            ?: schedulesCache.firstOrNull()
            ?: ScheduleConfig.defaultConfig()
    }

    suspend fun saveScheduleConfig(config: ScheduleConfig) = withContext(Dispatchers.Default) {
        val existing = database.schedulesQueries.getScheduleById(config.id).executeAsOneOrNull()
        val jsonStr = json.encodeToString(config)
        if (existing != null) {
            database.schedulesQueries.updateSchedule(jsonStr, config.id)
        } else {
            database.schedulesQueries.insertSchedule(config.id, jsonStr)
        }
        loadSchedulesCache()
    }

    suspend fun addSchedule(config: ScheduleConfig) = withContext(Dispatchers.Default) {
        database.schedulesQueries.insertSchedule(config.id, json.encodeToString(config))
        loadSchedulesCache()
    }

    suspend fun deleteSchedule(scheduleId: String) = withContext(Dispatchers.Default) {
        database.coursesQueries.deleteCoursesBySchedule(scheduleId)
        database.schedulesQueries.deleteSchedule(scheduleId)
        loadSchedulesCache()

        if (currentScheduleId == scheduleId && schedulesCache.isNotEmpty()) {
            switchSchedule(schedulesCache.first().id)
        }
    }

    // ==================== Courses ====================

    fun getCourses(): List<Course> = coursesCache

    suspend fun getCoursesAsync(scheduleId: String = currentScheduleId): List<Course> = withContext(Dispatchers.Default) {
        val rows = database.coursesQueries.getCoursesBySchedule(scheduleId).executeAsList()
        rows.map { row ->
            Course(
                id = row.id,
                name = row.name ?: "",
                teacher = row.teacher ?: "",
                location = row.location ?: "",
                startWeek = row.start_week?.toInt() ?: 1,
                endWeek = row.end_week?.toInt() ?: 20,
                dayOfWeek = row.day_of_week?.toInt() ?: 1,
                startSection = row.start_section?.toInt() ?: 1,
                endSection = row.end_section?.toInt() ?: 2,
                colorValue = row.color_value ?: 0xFF2196F3,
                weekType = WeekType.fromIndex(row.week_type?.toInt() ?: 0),
            )
        }
    }

    suspend fun addCourse(course: Course) = withContext(Dispatchers.Default) {
        database.coursesQueries.insertCourse(
            course.id,
            currentScheduleId,
            course.name,
            course.teacher,
            course.location,
            course.startWeek.toLong(),
            course.endWeek.toLong(),
            course.dayOfWeek.toLong(),
            course.startSection.toLong(),
            course.endSection.toLong(),
            course.colorValue,
            course.weekType.ordinal.toLong(),
        )
        loadCoursesCache()
    }

    suspend fun updateCourse(course: Course) = withContext(Dispatchers.Default) {
        database.coursesQueries.updateCourse(
            course.name,
            course.teacher,
            course.location,
            course.startWeek.toLong(),
            course.endWeek.toLong(),
            course.dayOfWeek.toLong(),
            course.startSection.toLong(),
            course.endSection.toLong(),
            course.colorValue,
            course.weekType.ordinal.toLong(),
            course.id,
        )
        loadCoursesCache()
    }

    suspend fun deleteCourse(courseId: String) = withContext(Dispatchers.Default) {
        database.coursesQueries.deleteCourse(courseId)
        loadCoursesCache()
    }

    fun hasConflict(course: Course, excludeId: String? = null): Boolean {
        return coursesCache.any { it.conflictsWith(course, excludeId = excludeId) }
    }

    // ==================== Clear ====================

    suspend fun clearAllCourseData() = withContext(Dispatchers.Default) {
        database.coursesQueries.deleteCoursesBySchedule(currentScheduleId)
        database.schedulesQueries.deleteSchedule(currentScheduleId)
        database.metadataQueries.deleteMetadata(KEY_CURRENT_SCHEDULE_ID)

        // Re-create default schedule
        val defaultConfig = ScheduleConfig.defaultConfig()
        database.schedulesQueries.insertSchedule(defaultConfig.id, json.encodeToString(defaultConfig))
        database.metadataQueries.setMetadata(KEY_CURRENT_SCHEDULE_ID, defaultConfig.id)
        currentScheduleId = defaultConfig.id

        loadSchedulesCache()
        loadCoursesCache()
    }

    companion object {
        private const val KEY_CURRENT_SCHEDULE_ID = "currentScheduleId"
    }
}
